rllib/evaluation/sampler.py

```diff
@@ -830,21 +834,24 @@ def _process_observations(
             hit_horizon = episode.length >= horizon and not dones[env_id]["__all__"]
             all_agents_done = True
             atari_metrics: List[RolloutMetrics] = _fetch_atari_metrics(base_env)
-            if atari_metrics is not None:
-                for m in atari_metrics:
-                    outputs.append(m._replace(custom_metrics=episode.custom_metrics))
-            else:
-                outputs.append(
-                    RolloutMetrics(
-                        episode.length,
-                        episode.total_reward,
-                        dict(episode.agent_rewards),
-                        episode.custom_metrics,
-                        {},
-                        episode.hist_data,
-                        episode.media,
+            if not episode_faulty:
+                if atari_metrics is not None:
+                    for m in atari_metrics:
+                        outputs.append(
+                            m._replace(custom_metrics=episode.custom_metrics)
+                        )
+                else:
+                    outputs.append(
+                        RolloutMetrics(
+                            episode.length,
+                            episode.total_reward,
+                            dict(episode.agent_rewards),
+                            episode.custom_metrics,
+                            {},
+                            episode.hist_data,
+                            episode.media,
+                        )
```

Comment:
do we need to clean up already collected metrics, partial data if episode_faulty is True?
actually I wonder if we can handle the faulty_episode case somewhere before all these, so we can assume that when we get here, the episode is good.
I need to go back and read sampler a bit more. am gonna work on this anyways.

Reply:
Yeah, was thinking about this, too. I think it's fine though:
* batch_mode=complete_episode: no problem here as episodes are discarded as a whole, if faulty
* batch_mode=truncate_episodes: when receiving the faulty flag, maybe we have already learned a bit (which is ok -> the beginning of the episode is most likely valid!); also we cannot undo the already (tune) reported results anyways; so unless the env is producing garbage rewards/data for a very long time *before* the failure, we shouldn't have a problem with learning.

Reply:
You can't reliably drop the episode in `batch_mode=truncated_episode`, b/c you might have already "shipped" some part of it to the learner for updating. Unless you want to retro-undo the already performed update step on the model's weights :)
All you can do really is to drop the current chunk of the episode (that has **not** "shipped" yet).

Reply:
In case: batch_mode=complete_episode, we drop this by saying: Don't add the built batch to `outputs` within sampler:

```
            # If, we are not allowed to pack the next episode into the same
            # SampleBatch (batch_mode=complete_episodes) -> Build the
            # MultiAgentBatch from a single episode and add it to "outputs".
            # Otherwise, just postprocess and continue collecting across
            # episodes.
            # If an episode was marked faulty, perform regular postprocessing
            # (to e.g. properly flush and clean up the SampleCollector's buffers),
            # but then discard the entire batch and don't return it.
            ma_sample_batch = sample_collector.postprocess_episode(
                episode,
                is_done=is_done or (hit_horizon and not soft_horizon),
                check_dones=check_dones,
                build=episode_faulty or not multiple_episodes_in_batch,
            )
            if not episode_faulty:
                if ma_sample_batch:
                    outputs.append(ma_sample_batch)

```